src/Castle.Core/Components.DictionaryAdapter/DictionaryAdapterFactory.cs

```diff
@@ -132,21 +132,32 @@ public DictionaryAdapterMeta GetAdapterMeta(Type type, DictionaryAdapterMeta oth
 
 			DictionaryAdapterMeta meta;
 
-			using (interfaceToMetaLock.ForReading())
+			interfaceToMetaLock.EnterReadLock();
+			try
 			{
 				if (interfaceToMeta.TryGetValue(type, out meta))
+				{
 					return meta;
+				}
+			}
+			finally
+			{
+				interfaceToMetaLock.ExitReadLock();
 			}
 
-			using (var heldLock = interfaceToMetaLock.ForReadingUpgradeable())
+			interfaceToMetaLock.EnterWriteLock();
```

Comment:
@stakx Just to confirm, we are happy with the performance hit here?

Reply:
@fir3pho3nixx - You know what, I think you are right. There is no real need to make this functional change now. Why "optimize" something when (a) I haven't established that this really *is* an optimization, and (b) when this is supposed to be a refactoring-only PR anyway?

I've reinstated the upgradeable read locks. After this PR, it should be easy to change things because the locking pattern that DynamicProxy will be present in only two locations (`SingletonDispenser` and `SynchronizedDictionary<,>`).